bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala

```diff
@@ -0,0 +1,168 @@
+      val tf  = TemplateFactoryImpl(templateDb)
+      val tfe = TemplateFolderExpansionFactory.expand(ss.folder, tf, true).unsafeGet
+      func(p, SkeletonStoreService.store(ss, tfe, db).program)
+    } catch {
+
+      // Because of the way specs works, any exception thrown in test construction will be reported
+      // as a missing constructor, which is confusing and unhelpful. So we'll spit out the
+      // exception here just so there's something useful to look at, and re-throw.
+      case t: Throwable =>
+        t.printStackTrace()
+        throw t
+
+    } finally {
+      db.getDBAdmin.shutdown()
+    }
+  }
+
+  /** Return the set of all integer library IDs that appear in the group. */
+  def libs(tg: ISPTemplateGroup): Set[Int] =
+    tg.getAllObservations
+      .asScala
+      .map(_.getDataObject.asInstanceOf[SPObservation])
+      .map(_.getLibraryId.parseInt)
+      .collect { case Success(n) => n }
+      .toSet
+
+  /** Return the list of all `ISPTemplateGroups`s in the given program. */
+  def groups(sp: ISPProgram): List[ISPTemplateGroup] =
+    sp.getTemplateFolder
+      .getTemplateGroups
+      .asScala
+      .toList
+
+  /** Assert that the given library IDs are included and excluded; call within a `should`. */
+  def checkLibs(label: String, tg: ISPTemplateGroup, incl: Set[Int], excl: Set[Int]) = {
+    val ls = libs(tg)
+    s"$label should include ${incl.mkString("{", ",", "}")} and exclude ${excl.mkString("{", ",", "}")}." in {
+      ls.filter(incl) == incl && ls.filter(excl).isEmpty
+    }
+  }
+
+  /** Retrieve the phase-2 targets from the given template group. */
+  def p2targets(g: ISPTemplateGroup): List[SPTarget] =
+    g.getTemplateParameters
+      .asScala.toList
+      .map(_.getDataObject.asInstanceOf[TemplateParameters])
+      .map(_.getTarget)
+
+  /** Retrieve the phase-2 magnitude in the given band, if any, from the given target. */
+  def p2mag(t: SPTarget, b: Band): Option[Double] =
+    Option(t.getTarget.getMagnitude(b).getOrNull).map(_.getBrightness)
+
+  /** True if a note with the given title exists at the root of the given template group. */
+  def existsNote(tg: ISPTemplateGroup, title: String) =
+    tg.getObsComponents
+      .asScala.toList
+      .map(_.getDataObject)
+      .collect { case n: SPNote => n }
+      .exists(_.getTitle == title)
+
+  /** Construct a phase-1 magnitude in the VEGA system. */
+  def p1Mag(n: Double, b: MagnitudeBand): Magnitude =
+    Magnitude(n, b, MagnitudeSystem.VEGA)
+
+  /** Construct an empty phase-1 sidereal target with the given magnitudes. */
+  def p1Target(ms: List[Magnitude]): SiderealTarget =
+    SiderealTarget.empty.copy(name = "test", magnitudes = ms)
+
+  /** Construct a phase-1 observation with the given blueprint and target. */
+  def p1Obs(bp: BlueprintBase, st: Target): Observation =
+    Observation.empty execState {
+      for {
+        _ <- blueprint := Some(bp)
```

Comment:
`Lens[A,B]` can also be thought of as a pair of operations `get: State[A,B]` and `set: B => State[A, Unit]` and the second is what the `:=` does ... it just lets you do a bunch of lens operations as a program in `State`, which is sometimes easier to follow. Ed Kmett's [lecture](https://www.youtube.com/watch?v=efv0SQNde5Q) here explains this style of lens if you're interested.